contracts/multisig-prover/src/encoding/mod.rs

```diff
@@ -1,4 +1,192 @@
-// Adding more encoding modules will require using feature flags in here to switch between them
-pub mod evm;
+pub mod abi;
 
-pub use evm::Data;
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{HexBinary, Uint256};
+use sha3::{Digest, Keccak256};
+
+use connection_router::msg::Message;
+use multisig::{key::Signature, msg::Signer};
+
+use crate::{
+    error::ContractError,
+    state::WorkerSet,
+    types::{BatchID, Command, CommandBatch, CommandType},
+};
+
+#[cw_serde]
+#[derive(Copy)]
+pub enum EncodingLanguage {
+    Abi,
+    Bcs,
+}
+
+fn make_command(msg: Message, encoding: EncodingLanguage) -> Result<Command, ContractError> {
+    Ok(Command {
+        ty: CommandType::ApproveContractCall, // TODO: this would change when other command types are supported
+        params: match encoding {
+            EncodingLanguage::Abi => abi::command_params(
+                msg.source_chain,
+                msg.source_address,
+                msg.destination_address,
+                msg.payload_hash,
+            )?,
+            EncodingLanguage::Bcs => todo!(),
+        },
+        id: command_id(msg.id),
+    })
+}
+
+fn make_transfer_operatorship(
+    worker_set: WorkerSet,
+    encoding: EncodingLanguage,
+) -> Result<Command, ContractError> {
+    let params = match encoding {
+        EncodingLanguage::Abi => abi::transfer_operatorship_params(&worker_set),
+        EncodingLanguage::Bcs => {
+            todo!()
+        }
+    }?;
+    Ok(Command {
+        ty: CommandType::TransferOperatorship,
+        params,
+        id: worker_set.hash(),
+    })
+}
+
+pub struct CommandBatchBuilder {
+    message_ids: Vec<String>,
+    new_worker_set: Option<WorkerSet>,
+    commands: Vec<Command>,
+    destination_chain_id: Uint256,
+    encoding: EncodingLanguage,
+}
+
+impl CommandBatchBuilder {
+    pub fn new(destination_chain_id: Uint256, encoding: EncodingLanguage) -> Self {
+        Self {
+            message_ids: vec![],
+            new_worker_set: None,
+            commands: vec![],
+            destination_chain_id,
+            encoding,
+        }
+    }
+
+    pub fn add_message(&mut self, msg: Message) -> Result<(), ContractError> {
+        self.message_ids.push(msg.id.clone());
+        self.commands.push(make_command(msg, self.encoding)?);
+        Ok(())
+    }
+
+    pub fn add_new_worker_set(&mut self, worker_set: WorkerSet) -> Result<(), ContractError> {
+        self.new_worker_set = Some(worker_set.clone());
```

Comment:
I still don't like it that worker set update is treated differently than messages, but I understand it's not what this PR is concerned about.